Move data that should be synced between computers to a common dir

Pentarctagon opened the pull request:
Fixes #8059

---

Still need to read through `preferences_list.hpp` again, but I don't recall seeing any that would require splitting the preferences file between synced and unsynced preferences.

The stuff moved under the `sync` folder was taken from https://github.com/wesnoth/wesnoth/issues/8059#issuecomment-2068444154

stevecotton commented on src/filesystem.cpp:
```diff
@@ -647,6 +648,10 @@ static void setup_user_data_dir()
 	create_directory_if_missing(get_logs_dir());
 
+	if(file_exists(get_unsynced_prefs_file()) && !file_exists(get_prefs_file())) {
+		copy_file(get_unsynced_prefs_file(), get_prefs_file());
```
Why copy instead of move?

Pentarctagon replied:
It could probably be a move, but just in case.

stevecotton commented on src/filesystem_common.cpp:
```diff
@@ -124,7 +124,12 @@ bool blacklist_pattern_list::match_dir(const std::string& name) const
 std::string get_prefs_file()
 {
-	return get_user_config_dir() + "/preferences";
+	return get_sync_dir() + "/preferences";
```
Not a request for a change, but making it a review comment will make Github create a discussion thread.

I looked at where the `user_config_dir` is still used, and it seems to only be used for the two below. It probably needs to stay for those, so can't be removed in this PR.
* remember the location of wesnothd.exe (lan_server.cfg)
* storing passwords (credentials-aes)

Pentarctagon replied:
I wasn't planning to move them, but also I'm not really sure how an encrypted credentials file counts as configuration in the first place. Then at that point it'd feel a bit silly to have just a single file (`lan_server.cfg`) put there.

CelticMinstrel commented:
> Still need to read through `preferences_list.hpp` again, but I don't recall seeing any that would require splitting the preferences file between synced and unsynced preferences.

preferences_list.hpp is incomplete, then. Off the top of my head, the most likely candidates for synching in preferences would be the achievements, the campaign completion status, and the input history for certain dialogs. On the other hand, anything related to screen resolution or scaling definitely should _not_ be synched. So, a split is definitely needed.

Pentarctagon commented:
I suppose. It seems like unwanted behavior in general if Wesnoth opens with a window that can't fit on the monitor it's being displayed on though.

Pentarctagon commented:
@CelticMinstrel should values set by the lua preferences API be synced? On one hand they're by definition tied to the add-on(s) that set them and add-ons aren't synced, but on the other hand it's pretty likely that someone has the same set (or at least a similar set) of add-ons installed on each of their computers.

For the mainline preferences, I'd say the following either make sense to keep synced or at least don't need to be kept unsynced:
* The MP alerts preferences (defined in preferences.hpp instead of preferences_list.hpp since I didn't think it made sense to unroll the loops they're used in.
* _last_cache_cleaned_ver
* addon_manager_saved_order_direction
* addon_manager_saved_order_name
* alias
* allow_observers
* ally_orb_color
* ally_sighted_interrupts
* auto_save_max
* blindfold_replay
* bell_volume
* campaign_server
* chat_lines
* chat_timestamp
* completed_campaigns
* confirm_end_turn
* custom_command
* delete_saves
* disable_auto_moves
* draw_delay
* editor_auto_update_transitions
* editor_draw_hex_coordinates
* editor_draw_num_of_bitmaps
* editor_draw_terrain_codes
* enable_planning_mode_on_start
* encountered_terrain_list
* encountered_units
* enemy_orb_color
* fi_blocked_in_game
* fi_friends_in_game
* fi_invert
* fi_vacant_slots
* floating_labels
* fullscreen
* grid
* hide_whiteboard
* history
* host
* idle_anim
* idle_anim_rate
* lobby_joins
* lobby_whisper_friends_only
* locale
* login
* maximized
* message_bell
* minimap_draw_terrain
* minimap_draw_units
* minimap_draw_villages
* minimap_movement_coding
* minimap_terrain_coding
* moved_orb_color
* mp_countdown
* mp_countdown_action_bonus
* mp_countdown_init_time
* mp_countdown_reservoir_time
* mp_countdown_turn_bonus
* mp_fog
* mp_level_type
* mp_random_start_time
* mp_server_warning_disabled
* mp_shroud
* mp_turns

And these make sense to not keep synced:
* auto_pixel_scale - related to the monitor size
* core - in case it ends up set to a UMC core which hasn't been installed
* dir_bookmarks - related to the local filesystem structure
* draw_delay - used by the FPS limiter
* editor_chosen_addon - add-ons aren't synced
* editor_recent_files - might contain files that don't exist on all computers its synced to
* gui2_theme - IIRC it was said this effectively isn't used currently, but just in case
* mp_era - era might not be installed
* mp_level - scenario might not be installed
* mp_modifications - modifications might not be installed
* mp_server_program_name - might be different across computers, particularly operating systems
* 

Also I think I'm going to just remove editor_default_dir since it's not actually used. Might also take another shot and getting the MP alerts preferences to work as part of preferences_list.hpp.

CelticMinstrel commented:
> I suppose. It seems like unwanted behavior in general if Wesnoth opens with a window that can't fit on the monitor it's being displayed on though.

???

That's exactly why resolution information should _not_ be synched.

> @CelticMinstrel should values set by the lua preferences API be synced? On one hand they're by definition tied to the add-on(s) that set them and add-ons aren't synced, but on the other hand it's pretty likely that someone has the same set (or at least a similar set) of add-ons installed on each of their computers.

There are no add-ons that use the Lua preferences API, and I don't expect there ever would be. The API was added for use by integration testing and is part of the plugins API. That means it might also be used for things like bots, but wouldn't appear in an add-on. I would say it should sync the values if the API sets a syncable preference, and not sync them otherwise. In other words, the fact that it was the Lua API that set the preference shouldn't have any impact.

Most of your split seems to make sense for me, with a few exceptions. The following seem like maybe they shouldn't be synched:

> * bell_volume
> * fullscreen
> * login - only because I'm not sure what it is
> * music_volume
> * remember_password - because the remembered password itself is not synched
> * sound_volume
> * ui_volume

I'm not really sure about whether the volume controls should be synched or not. The fullscreen setting _might_ be safe to sync, but for consistency we probably shouldn't.

I think these ones from your "don't sync" list might be fine to sync:

> * mp_era
> * mp_level
> * mp_modifications
> * selected_achievement_group
> * sp_modifications

These are basically just remembering your last selection, and they already handle the case where the last selection doesn't exist (maybe you uninstalled that add-on). So while it could make sense to omit them from the sync group, it should also be fine to include them.

Also, I think you missed achievements from your list? It looks like you do have completed campaigns (the state of the laurels and everything). Are you also missing the GUI2 input history? I see "history" on the list but I'm not sure if that's the same thing.

Pentarctagon commented:
> login - only because I'm not sure what it is

It's the username most recently used for logging into the multiplayer server.

> remember_password - because the remembered password itself is not synched

I'm not sure how that's relevant. That checkbox is checked on the UI when someone doesn't want to need to enter their password each time. That the password may not exist doesn't change that they don't want to need to enter it again the next time.

Updated other than those two.

> Also, I think you missed achievements from your list? It looks like you do have completed campaigns (the state of the laurels and everything). Are you also missing the GUI2 input history? I see "history" on the list but I'm not sure if that's the same thing.

Apparently I somehow missed adding achievements of all things to the `prefs_list`, so that's added now. `history` is the GUI2 input history.

soliton- commented:
"remove editor_default_dir as it's not used" is a confusing commit since it appears to remove the use of editor_default_dir.

soliton- commented on src/mp_ui_alerts.hpp:
```diff
@@ -42,7 +42,4 @@ namespace mp::ui_alerts {
 	bool get_def_pref_notif(const std::string &);
 	bool get_def_pref_lobby(const std::string &);
-
-	// Note, this list of items must match those ids defined in data/gui/dialogs/mp_alerts_options.cfg
```
This seems like a useful comment that should rather be updated and not removed.

Pentarctagon replied:
I'll re-add it.

soliton- commented on src/gui/dialogs/multiplayer/mp_alerts_options.cpp:
```diff
@@ -34,6 +34,19 @@
 namespace gui2::dialogs
 {
 
+const std::vector<std::string> items{
```
Since the namespace doesn't give context anymore for this variable a better name would be good. Perhaps `mp_ui_alerts`.

Pentarctagon replied:
Well, it doesn't exist anymore either.

soliton- replied:
I was looking at commit-by-commit. Is the intention to squash this PR?

soliton- replied:
IMO at least the reorganization of the MP alerts options and preferences and the removal of the editor_default_dir preference should be their own commits.

Pentarctagon replied:
I will try to clean it up like that. Previous times when I've tried to move around commits that touch the same code in a PR though, I've just made a mess of things.

soliton- commented on src/filesystem_common.cpp:
```diff
@@ -124,7 +124,12 @@ bool blacklist_pattern_list::match_dir(const std::string& name) const
 
 std::string get_prefs_file()
 {
-	return get_user_config_dir() + "/preferences";
```
Why is this changed?

Pentarctagon replied:
Because the user config dir is not what will be synced.

soliton- replied:
But the preferences file will continue to exist, right? You're just adding another preferences file in the sync dir?

Pentarctagon replied:
It doesn't make sense to me to have the synced and unsynced preferences saved to entirely different locations though.

Pentarctagon replied:
I can move the unsynced preferences back if you prefer.

Correct, syncing will not work. Steam expects the userdata to be in a specific place for syncing to work today as well - if someone moves it somewhere else, then their data will no longer be synced.

Pentarctagon replied:
There would also need to be some sort of handling for unknown preferences added as well, since at least on Linux the config dir tends to not be versioned so with the current implementation in this PR it would result in unsynced preferences added to the dev version being lost when someone switches between the stable and dev versions.

soliton- replied:
1. I would keep `get_prefs_file()` as the function to return the normal preferences file.
2. I would either properly use the user config dir (leave the preferences there and put the sync dir there as well) or remove support for it.

Pentarctagon replied:
I don't know what you mean by "put the sync dir there as well" - only a single directory gets synced by Steam, it doesn't accomplish anything to potentially have a sync dir in two places. I will remove the config dir functionality in a separate PR though.

soliton- replied:
The sync dir should be where the preference file is. The preference file is in the user config dir at the moment. So logically the sync dir would be in the user config dir. No two locations, just one.

Since you appear to not like the user config dir the other option is to remove support for it and move everything to the user data dir. A mix of both makes little sense.

stevecotton replied:
Why should the sync dir be where the preference file is? From your message, it's not clear what your reasoning is.

Pentarctagon replied:
This makes no sense at all. Currently what's synced is the saves folder, not the preferences folder, and it would be absolutely bizarre to move everything that is going to be synced to the user config directory instead.

soliton- replied:
Sorry must have missed the notification for this.

Looks like I got confused that this is only about preferences. So it sounds like the only option is to remove the user config dir. If that is not supposed to happen in this PR then the preference file should not move in this PR either.

stevecotton replied:
Again, what's your reasoning? Why are those the two options?

soliton- replied:
What other options are there? The preference file should move to the user data dir so it is next to the sync dir. It makes no sense to have a config dir with no config in it. No two options left. Just one.

Pentarctagon replied:
I intend to do this and figure out how to get the assert discussed in https://github.com/wesnoth/wesnoth/pull/9016#discussion_r1650036348 working in a followup PR. The preferences location also already defaults to the userdata dir as of #8848 anyway - all this PR does is split it into two files and and move one file into a subdirectory.

soliton- replied:
As far as I can tell this PR currently changes `get_prefs_file` to return the path to the new synced preferences and adds the function `get_unsynced_prefs_file` to return the path to the preference file but from the user data dir.

I would keep `get_prefs_file` as is since there is no intention to remove the user config dir in this PR and add a `get_synced_prefs_file` to return the path to the new synced preferences file.

Pentarctagon replied:
I've renamed it to `get_synced_prefs_file`, that way it's clear what each is meant to do.

Pentarctagon commented:
> "remove editor_default_dir as it's not used" is a confusing commit since it appears to remove the use of editor_default_dir.

How can it be used when nothing ever sets it to a value?

Pentarctagon commented on src/preferences/preferences.cpp:
```diff
@@ -259,10 +259,6 @@ void prefs::load_preferences()
 
-	// We save the password encrypted now. Erase any saved passwords in the prefs file.
-	erase("password");
-	erase("password_is_wrapped");
```
The next commit to write out the synced vs unsynced preferences will no longer write out unknown preferences. They'd also need to load a preferences file from 1.12 or earlier for it to exist at all.

soliton- replied:
As you mentioned elsewhere it is probably not really workable to remove unknown preferences or you will lose data whenever you run wesnoth with a newer/older preferences file.

Pentarctagon replied:
Then should unknown preferences be synced or unsynced?

soliton- replied:
Ideally wherever they came from. If they came from the unsynced file then they should be written back to it as well.

Pentarctagon replied:
Pushed a commit to do this, though I haven't tested it yet.

Pentarctagon replied:
Tested and it appears to work as intended.

soliton- commented:
> How can it be used when nothing ever sets it to a value?

You can edit the preferences file.

Pentarctagon commented:
How would anyone know it even exists?

Pentarctagon commented:
Yes, that's essentially what I meant. No *person* is using the preference given they essentially need to read through the source code, find it by accident, and then go set it manually in the preferences file.

Pentarctagon commented:
I probably should've marked this as a draft originally given how incomplete it ended up being when I originally opened this PR. In any case though, I think the code is essentially complete aside from addressing review comments.

CelticMinstrel commented on src/preferences/preferences.cpp:
```diff
@@ -309,24 +315,64 @@ void prefs::load_preferences()
+	for(const char* attr : synced_children_) {
+		for(const auto& child : preferences_.child_range(attr)) {
+			config& ch = synced.add_child(attr);
+			ch.append_children(child);
```
Is this correct…? Something seems wrong about this. It adds a child tag to the synched prefs with the appropriate name, and then appends children from the child tag to that new tag… but how can that possibly be correct?

Pentarctagon replied:
When I didn't do that, it was dropping the `attr` child tag and its attributes (tested with the completed campaigns tags).

Pentarctagon replied:
More specifically, taking the example from the comments in `preferences.cpp`:
```
completed_campaigns = "A,B,C"
[completed_campaigns]
	[campaign]
		name = "A"
		difficulty_levels = "EASY,MEDIUM"
	[/campaign]
[/completed_campaigns]
```
it saves it as:
```
[completed_campaigns]
	[campaign]
		difficulty_levels="EASY,MEDIUM"
		name="A"
	[/campaign]
	[campaign]
		difficulty_levels=""
		name="B"
	[/campaign]
	[campaign]
		difficulty_levels=""
		name="C"
	[/campaign]
[/completed_campaigns]
```
and maintains it in that form no matter how many times I resave it. If I instead change it to `synced.append_children(child);` then it gets saved as:
```
[campaign]
	difficulty_levels="EASY,MEDIUM"
	name="A"
[/campaign]
[campaign]
	difficulty_levels=""
	name="B"
[/campaign]
[campaign]
	difficulty_levels=""
	name="C"
[/campaign]
```
after the first save and then removed entirely after the second save.

Pentarctagon replied:
Apparently this still doesn't preserve any attributes under the outermost tag either.

CelticMinstrel replied:
What exactly is wrong with just `synced.add_child(attr, child)`?

Pentarctagon replied:
That works, apparently.

Pentarctagon commented:
Barring any additional unresolved comments, I intend to merge this a week after #9016

soliton- commented on src/filesystem_common.cpp:
```diff
@@ -122,9 +122,14 @@ bool blacklist_pattern_list::match_dir(const std::string& name) const
+
+std::string get_unsynced_prefs_file()
+{
+	return get_user_data_dir() + "/preferences";
```
```suggestion
	return get_config_data_dir() + "/preferences";
```
It makes no sense to just change this location. Either dissolve the user config dir or keep it as is.

Pentarctagon replied:
I've already said how I intend to handle this multiple times now. I don't have anything else to say about it.

soliton- replied:
So this is intentionally doing part of the changes of a later PR? What's the point of that? I do not understand the insistence on doing changes in a confusing way. Even changes that affect people working on wesnoth by moving important files around without any mention in the commit message or PR description.

I'm _very_ opposed to this. I read commits all the time. I do want wesnoth's commit history to be as easy to read and follow as possible. I'm happy to do whatever changes myself if this is about being more work or something.

Pentarctagon replied:
There, done.

soliton- commented on src/desktop/paths.cpp:
```diff
@@ -200,7 +200,7 @@ std::vector<path_info> game_paths(std::set<GAME_PATH_TYPES> paths)
 	static const std::string& game_data_dir = pretty_path(game_config::path);
 	static const std::string& game_user_data_dir = pretty_path(filesystem::get_user_data_dir());
-	static const std::string& game_user_pref_dir = pretty_path(filesystem::get_user_config_dir());
+	static const std::string& game_user_pref_dir = pretty_path(filesystem::get_user_data_dir());
```
I don't see a reason to keep this.

Pentarctagon replied:
Removed.

stevecotton commented:
`git grep config-path` still lists doc/man/wesnoth.6

Please add a changelog entry too.

Pentarctagon commented:
Done.
